### PR TITLE
Discourage usage of global variables

### PR DESCRIPTION
**Pull request by LinqLover**
Global variables should be used as seldom as possible; their usage is not considered idiomatic for the most time. Emphasize this aspect in the relevant sections. Nuke description of Morphic globals (`World` and `ActiveHand`) that are meant to be deprecated and should not accessed manually at all. See also: [[squeak-dev] Changeset: Eliminating global state from Morphic](http://forum.world.st/Changeset-Eliminating-global-state-from-Morphic-td5121690.html)

### CODE REVIEW

File: Model/Model.tex [hunk 1]
```diff
@@ -858,17 +858,10 @@ \subsubsection{Other useful global variables}
 (Smalltalk at: #Smalltalk) == Smalltalk --> true
 \end{code}
 
+% ct: TODO DISCUSS: Shall we really recommend the usage of \ct{Sensor} any longer?}
```
Comment (LinqLover) [inline Model/Model.tex:861]: @codeZeilen I wonder for the other globals listed here, `Undeclared`, `SystemOrganization`, and maybe even `Transcript`, whether we should recommend their usage at all? There are `Smalltalk undeclared`, `SystemOrganizer default`, and `Project current transcript` that possibly are more idiomatic ways. Can we drop these descriptions, too? And if yes, should we explain the relevant concepts somewhere else?

Comment (codeZeilen) [reply Model/Model.tex:861]: We definitely want to describe them somewhere else, including World and ActiveHand, they are all useful. The whole section has to be re-written to introduce the notion of global variables, and then explain why they are problematic, and then to introduce the idiomatic ways for all the objects in the old list.

**Review by codeZeilen — CHANGES_REQUESTED** [review]
We have to rework the section in order to solve that problem the right way.

File: Model/Model.tex (hunk 1 above)
Comment (LinqLover) [reply Model/Model.tex:861]: Later in this section, there is already a paragraph discouraging the use of global variables, but it's probably not precise enough, see <https://github.com/codeZeilen/SqueakByExample-english/blob/b8470833881043eafd33f06de62b51fee38e1313/Model/Model.tex#L880>.

World, ActiveHand, and Sensor are all Morphic-specific. Would the new description of `#currentWorld` and `#currentHand` maybe belong there?

**Comment by codeZeilen** [comment]
Yes, I saw that paragraph and I think that the location could work for the
discussion. I will try in the next few weeks to write it. If you feel like
it and are willing to endure the review cycles you can also give it a try.
:)


Am Fr., 25. Sept. 2020 um 17:55 Uhr schrieb Christoph Thiede <
notifications@github.com>:

**Comment by LinqLover** [comment]
> If you feel like it and are willing to endure the review cycles you can also give it a try. :)

I'm trying my luck now :-)

**Comment by LinqLover** [comment]
@codeZeilen Hmm, didn't the "request review" button trigger a notification, or didn't you have time yet? 😅 

image

File: Environment/Environment.tex [hunk 2]
```diff
@@ -516,6 +516,13 @@ \subsection{Browsing programmatically}
 To learn more about what is available, explore the class \ct{SystemNavigation} with the browser.
 Further navigation examples can be found in the FAQ (\appref{faq}).
 
+Beside \ct{SystemNavigation}, there are also several subclasses of \clsind{Categorizer} that manage the categories in which classes or methods are organized.
```
Comment (codeZeilen) [inline Environment/Environment.tex:519]: beside -> besides

File: Environment/Environment.tex [hunk 3]
```diff
@@ -516,6 +516,13 @@ \subsection{Browsing programmatically}
...
+Concretely, a \clsind{ClassOrganizer} records the organization of methods in a class, and analogously, a \clsind{SystemOrganizer} records the organization of classes within system categories.
```
Comment (codeZeilen) [inline Environment/Environment.tex:520]: Concretely, -> In particular,

File: Environment/Environment.tex [hunk 4]
```diff
@@ -516,6 +516,13 @@ \subsection{Browsing programmatically}
...
+For example, you can retrieve all classes that are in the system category \ct{Collections-Strings} by sending \mthind{SystemOrganizer}{allClassesIn:} to the default \ct{SystemOrganizer} instance:
```
Comment (codeZeilen) [inline Environment/Environment.tex:521]: Text und Code passen nichts zueinander. Text sagt "allClassesIn:", der Code verwendet "classesIn."

File: Model/Model.tex [hunk 5]
```diff
@@ -826,11 +826,21 @@ \subsection{Global variables}
 Global variables are accessible everywhere. 
 Every class is named by a global variable; in addition, a few globals are used to name special or commonly useful objects.
 
-The variable \glbind{Transcript} names an instance of \clsind{TranscriptStream}, a stream that writes to a scrolling window.
-The following code displays some information and then goes to the next line in the \ct{Transcript}.
+However, what may sound so handy and convenient at the first glance has turned out as bad practice during the evolution of \sq.
+While we state in rule 4 that ``everything happens by message sends'', the concept of globals is a violation of this rule.
+Accessing a global allows you to exchange information between two objects\,---\,\emph{without} sending any message.
+As a consequence, when using globals, we loose an important advantage of the \st object model, which is the total encapsulation of every object's state.
```
Comment (codeZeilen) [inline Model/Model.tex:832]: without "total"

File: Model/Model.tex [hunk 6]
```diff
@@ -826,11 +826,21 @@ \subsection{Global variables}
...
+Also, global variables introduce \emph{global state} as mentioned earlier which impedes the modularity of \st objects.
```
Comment (codeZeilen) [inline Model/Model.tex:833]: maybe add something like "... and hinders us to use a \st object directly without taking care of the context."

"context" does not really sound right yet, but something that describes the actual consequence for users would be good here. "Modularity" is a rather abstract thing.

File: Model/Model.tex [hunk 7]
```diff
@@ -826,11 +826,21 @@ \subsection{Global variables}
...
+
+For these reasons, current practice is to strictly limit the use of global variables; it is usually better to use class instance variables or class variables and to provide class methods to access them.
+Indeed, if \squeak were to be implemented from scratch today, most of the global variables that are not classes would be replaced by singletons.
```
Comment (codeZeilen) [inline Model/Model.tex:836]: I would remove that sentence. The discussion of things that could have been done differently is not important for beginners. They might just wonder: Why is it still in there then?

File: Model/Model.tex [hunk 8]
```diff
@@ -826,11 +826,21 @@ \subsection{Global variables}
...
+Still, there are a number of globals defined in \sq, but for most of them, there exists a preferred alternative too, which is a message that usually can be sent to a particular class.
```
Comment (codeZeilen) [inline Model/Model.tex:837]: "exists" -> "is" (easier to read)

File: Model/Model.tex [hunk 9]
```diff
@@ -839,11 +849,11 @@ \subsection{Global variables}
...
 So, if you experience some sluggishness and are writing to the Transcript, think about collapsing it.}
 
-\subsubsection{Other useful global variables}
+Transcript outputs are mostly used for debugging and experimenting purposes where code quality is not the most important thing.
+If you want to use a transcript in a real application, you should use \ct{Project current transcript} instead, which makes sure that the transcript you are using actually matches your current \sq environment.
 
-\begin{itemize}
-\item
-\ct{Smalltalk} is the instance of \ct{SystemDictionary} that defines all of the globals\,---\,including \ct{Smalltalk} itself.   
+Another useful global variable is \ct{Smalltalk} which holds the instance of \ct{SystemDictionary} that defines all of the globals\,---\,including \ct{Smalltalk} itself.
```
Comment (codeZeilen) [inline Model/Model.tex:855]: You are obviously more up-to-date but I thought that the direct use of `Smalltalk` was also discouraged?

File: Morphic/Morphic.tex [hunk 10]
```diff
@@ -559,6 +559,31 @@ \subsection{Keyboard events}
 To discover the key values, you can open a Transcript window and add \glbind{Transcript} \ct{show: anEvent keyValue} to \mthref{handleKeystroke}.
 The \ct{anEvent} argument of \ct{handleKeystroke:} is an instance of \clsind{KeyboardEvent}, another subclass of \clsind{MorphicEvent}. Browse this class to learn more about keyboard events.
 
+\subsection{Global accessors}
```
Comment (codeZeilen) [inline Morphic/Morphic.tex:562]: Maybe a more detailed heading could be "Global accessors for user interface objects"?

File: Morphic/Morphic.tex [hunk 11]
```diff
@@ -559,6 +559,31 @@ \subsection{Keyboard events}
...
+
+In complex applications, one may need to refer to the currently processed event without having it available in the current method.
+While we should avoid such situations in general, there is an accessor message defined on \ct{Object}, \mthind{Object}{currentEvent}, that allows us to retrieve the event currently being handled again from any method in any object.
+To do so, we can simply write \ct{self currentEvent} in an arbitrary method.
+Analogously, there are \mthind{Object}{currentHand} and \mthind{Object}{currentWorld} to retrieve the current \clsind{HandMorph} (which represents your cursor) or \clsind{PasteUpMorph} (which contains all other morphs displayed at your screen).
+
+In some cases, one also may want to directly control the user interaction without following the usual event handling control flow.
```
Comment (codeZeilen) [inline Morphic/Morphic.tex:569]: "In some cases, one may also want to..."

File: Morphic/Morphic.tex [hunk 12]
```diff
@@ -559,6 +559,31 @@ \subsection{Keyboard events}
...
+As an example, we want to track the mouse cursor and find out which point the user clicks next.
```
Comment (codeZeilen) [inline Morphic/Morphic.tex:570]: "As an example, we may want to..." so the sentence parallels with the sentence before.

File: Morphic/Morphic.tex [hunk 13]
```diff
@@ -559,6 +559,31 @@ \subsection{Keyboard events}
...
+Implementing this via the usual event-handling protocol would be complicated because we want to handle clicks that are not made to a morph controlled by ourselves.%
```
Comment (codeZeilen) [inline Morphic/Morphic.tex:571]: maybe: "we want to handle click events for morphs whose methods we do not control."

File: Morphic/Morphic.tex [hunk 14]
```diff
@@ -559,6 +559,31 @@ \subsection{Keyboard events}
...
+\footnote{Actually, there is an advanced mechanism called \emph{event-bubbling and -capturing filters} that would be applicable to this problem.
+	However, this approach requires a bit more code and thus is not optimally suited for our simple use case.}
+For cases like this, there is the class \clsind{EventSensor} an instance of which you can retrieve by evaluating \ct{EventSensor} default.
```
Comment (codeZeilen) [inline Morphic/Morphic.tex:574]: shouldnt this by \mthind or some inline code format?

File: Morphic/Morphic.tex [hunk 15]
```diff
@@ -559,6 +559,31 @@ \subsection{Keyboard events}
...
+To implement our example, we can do the following:
+\begin{method}{Retrieving information about the cursor from \ct{EventSensor} without using events.}
+getPositionFromClick
+	| position |
+	[EventSensor default anyButtonPressed] whileFalse: [
+		position := EventSensor default cursorPoint].
+	^ position
+\end{method}
+
+However, keep in mind that all these approaches use global state which can lead to problems in particular situations (just imagine a \sq image that deals with multiple cursors\footnote{This is actually the case in some multi-touch implementations for \sq{}.}, which cursor's position are we retrieving now?).
```
Comment (codeZeilen) [inline Morphic/Morphic.tex:584]: ~~in particular situations~~

**Review by codeZeilen — CHANGES_REQUESTED** [review]

File: Model/Model.tex (hunk 6 above)
Comment (LinqLover) [reply Model/Model.tex:833]: What about "environment" instead of "context"?

File: Model/Model.tex (hunk 9 above)
Comment (LinqLover) [reply Model/Model.tex:855]: Not sure about this, at least it has more than 1000 senders in my image. @marceltaeumel Sorry for pinging, is it best practice to use `Smalltalk` directly or `SmalltalkImage current` instead?

File: Environment/Environment.tex (hunk 2 above)
Comment (LinqLover) [reply Environment/Environment.tex]: 54bf6db8700fd02c4d21d8f3af08e68843378390

File: Environment/Environment.tex (hunk 3 above)
Comment (LinqLover) [reply Environment/Environment.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390

File: Environment/Environment.tex (hunk 4 above)
Comment (LinqLover) [reply Environment/Environment.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390

File: Model/Model.tex (hunk 5 above)
Comment (LinqLover) [reply Model/Model.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390

File: Model/Model.tex (hunk 7 above)
Comment (LinqLover) [reply Model/Model.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390

File: Model/Model.tex (hunk 8 above)
Comment (LinqLover) [reply Model/Model.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390

File: Morphic/Morphic.tex (hunk 10 above)
Comment (LinqLover) [reply Morphic/Morphic.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390

File: Morphic/Morphic.tex (hunk 11 above)
Comment (LinqLover) [reply Morphic/Morphic.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390.

File: Morphic/Morphic.tex (hunk 12 above)
Comment (LinqLover) [reply Morphic/Morphic.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390.

File: Morphic/Morphic.tex (hunk 13 above)
Comment (LinqLover) [reply Morphic/Morphic.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390.

File: Morphic/Morphic.tex (hunk 14 above)
Comment (LinqLover) [reply Morphic/Morphic.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390.

File: Morphic/Morphic.tex (hunk 15 above)
Comment (LinqLover) [reply Morphic/Morphic.tex]: Resolved via 54bf6db8700fd02c4d21d8f3af08e68843378390.

**Comment by LinqLover** [comment]
@codeZeilen Thanks for the detailed review! Two open questions are left. :-)

File: Model/Model.tex (hunk 6 above)
Comment (codeZeilen) [reply Model/Model.tex]: Yes, "environment" would work :)

File: Model/Model.tex (hunk 9 above)
Comment (codeZeilen) [reply Model/Model.tex]: wow, ok nevermind :) I guess it depends on whether one wants to access information on the execution environment (e.g. `Smalltalk isBigEndian`) or whether one wants to access the environment.

**Comment by LinqLover** [comment]
@codeZeilen Looks like there are no open change requests, anything left to do, or can you merge this?

**Review by codeZeilen — APPROVED** [review]

**Pull request merged**